sdks/python/apache_beam/runners/dataflow/dataflow_runner.py

```diff
@@ -360,6 +360,16 @@ def visit_transform(self, transform_node):
 
   def run_pipeline(self, pipeline, options):
     """Remotely executes entire pipeline or parts reachable from node."""
+    # Label goog-dataflow-notebook if pipeline is initiated from interactive
+    # runner.
+    if pipeline.interactive:
```

Comment:
The change could be limited to:
- here detect, whether we are in interactive environment or not. (For example, check whether certain imports are loaded?) Could also be a utility method somewhere, to check whether this in an interactive method.
- If yes, add the labels.

Reply:
I see your point! Yes, I have the [capability](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/interactive/interactive_environment.py#L131) to check if current interpreted code is in a notebook or not. This branch will need a rebase against master to take those changes.

To clartify the process:
When a DataflowRunner tries to run a job from a given pipeline,
1. Check if the module `interactive_environment` is imported by checking the `sys.modules` dictionary;
2. Check if `current_env().is_in_notebook`;
3. If yes, label the job.

Reply:
Could we move https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/interactive/interactive_environment.py#L100:L102 to a common utility function, and each runner if they want could call this without worrying about require additional imports?

Reply:
Discussed with David and Sam. Since we also want to track jobs started from notebook even if the user never uses `InteractiveRunner`, checking the environment might just be the only way to do it.
By putting the logic into a try-except block as it is, we could avoid introducing `ipython` dependency into `DataflowRunner`. If the `[interactive]` dependency is never installed and current execution_path has never imported `ipython`, the code would just never be executed.

I'll move the logic into a standalone utility module and import it in DataflowRunner to do the check.